tap_eip712_message/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Eip712Error {
+    /// `alloy` wallet error
+    #[error(transparent)]
+    WalletError(#[from] alloy::signers::Error),
+
+    /// `alloy` wallet error
+    #[error(transparent)]
+    SignatureError(#[from] alloy::primitives::SignatureError),
+
+    /// Error when signature verification fails
+    #[error("Expected address {expected} but received {received}")]
+    VerificationFailed {
+        expected: Address,
+        received: Address,
+    },
+}
+
+/// EIP712 signed message
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct EIP712SignedMessage<M: SolStruct> {
+    /// Message to be signed
+    pub message: M,
+    /// ECDSA Signature of eip712 hash of message
+    pub signature: Signature,
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SignatureBytes([u8; 65]);
+
+pub trait SignatureBytesExt {
+    fn get_signature_bytes(&self) -> SignatureBytes;
+}
+
+impl SignatureBytesExt for Signature {
+    fn get_signature_bytes(&self) -> SignatureBytes {
+        SignatureBytes(self.as_bytes())
+    }
+}
+
+/// Unique identifier for a message
+///
+/// This is equal to the hash of the contents of a message, excluding the signature.
+/// This means that two receipts signed by two different signers will have the same id.
+///
+///
+/// This cannot be used as a unique identifier for a message, but can be used as a key
+/// for a hashmap where the value is the message.
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct MessageId(pub [u8; 32]);
+
+impl<M: SolStruct> EIP712SignedMessage<M> {
+    /// Creates a signed message with signed EIP712 hash of `message` using `signing_wallet`
+    pub fn new(
+        domain_separator: &Eip712Domain,
+        message: M,
+        signing_wallet: &PrivateKeySigner,
+    ) -> Result<Self, Eip712Error> {
+        let recovery_message_hash = message.eip712_signing_hash(domain_separator);
+
+        let signature = signing_wallet.sign_hash_sync(&recovery_message_hash)?;
+
+        Ok(Self { message, signature })
+    }
+
+    /// Recovers and returns the signer of the message from the signature.
+    pub fn recover_signer(&self, domain_separator: &Eip712Domain) -> Result<Address, Eip712Error> {
+        let recovery_message_hash = self.message.eip712_signing_hash(domain_separator);
+        let recovered_address = self
+            .signature
+            .recover_address_from_prehash(&recovery_message_hash)?;
+        Ok(recovered_address)
+    }
+
+    /// Checks that receipts signature is valid for given verifying key, returns `Ok` if it is valid.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`crate::Error::SignatureError`] if the recovered address from the
+    /// signature is not equal to `expected_address`
+    ///
+    pub fn verify(
+        &self,
+        domain_separator: &Eip712Domain,
+        expected_address: Address,
+    ) -> Result<(), Eip712Error> {
```

Comment:
I was curious why you didn't think about returning a `bool` here. Returning `Err` here doesn't just mean the address didn't match.

Reply:
Signature recovery can fail somehow. I think it's because it can fail converting the signature into the recovery bits.

Reply:
It happens a lot when there's a fail in the ser/de process

Reply:
That's what I mean. So it should be Err if that happens, then Ok(true/false) depending on the check result?

Reply:
I don't have strong opinions, do you think I should update this?

Reply:
I'm trying to understand, is it a system error if the address doesn't match?

Reply:
It was made this way because it's easier to just match on all the errors, and one of them is "VerificationError". But we can change, no problems.

Reply:
Is this necessary for the PR approval?

Reply:
Is it a system error if there isn't a match though? Just trying to understand.

Reply:
It's used only here https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/1fc51a3ff9ce74027b47a0e6d026a5bedd9ca00c/tap_core/src/manager/adapters/signature.rs#L28

Reply:
Where in the code do you match on it?

Reply:
https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/1fc51a3ff9ce74027b47a0e6d026a5bedd9ca00c/tap_core/src/manager/adapters/signature.rs#L34-L44

Reply:
That seems to where you initialize the error. You mentioned matching on the error types in a comment above so I was wondering if that actually happens anywhere for this error?

Reply:
Sorry, I thought it was the inverse. These are legacy decisions carried over with the refactors.

Reply:
Do you see what's confusing about the error handling here? 

```rust
if self 
     .verify_signer(recovered_address) 
     .await 
     .map_err(|e| Error::FailedToVerifySigner(e.to_string()))? 
 { 
     Ok(()) 
 } else { 
     Err(Error::InvalidRecoveredSigner { 
         address: recovered_address, 
     }) 
 } 
```

Reply:
Returning bool for this method